charts/k8ssandra/values.yaml

```diff
@@ -74,9 +74,9 @@ cassandra:
   allowMultipleNodesPerWorker: false
 
   # -- Cluster-level heap configuration
-  heap:
-   size:
-   newGenSize:
+  #heap:
```

Comment:
Interesting, I thought we checked to see if the value was set and calculated if not. It could be we check to see if it is defined, not empty. In either case, I like the inclusion of the keys here as it drives the reference docs (see the regenerated files from the GH action). If we don't want to leverage empty keys here we should update the inline docs to let the user know which keys are available (again, reference the previously generated docs).

Reply:
@bradfordcp right, the calculated heap feature is something post 1.0 as there are some thoughts about including an up-front `automated survey` of the configuration, which _should_ provide a more accurate calculated set of heap defaults. 

For `v1.0`, we are planning to stay consistent with no empty keye.  

I'll update the inline documentation to reference the `k8ssandra.io` docs.  If not specified in the `values.yaml`, 
the recommended default values reference the cass-operator's [jvm_server_configuration](https://github.com/datastax/cass-operator/blob/master/docs/user/jvm_server_configuration.md).

@jsanda - keep me honest here based on our past discussions in Slack, etc. :smile:

Reply:
Added additional documentation reference as noted in the comment.

Reply:
Am I understanding correctly that, due to these changes, our automatically generated settings docs will not contain specific entires for any of `cassandra.heap`, `cassandra.heap.size`, `cassandra.heap.newGenSize`, `cassandra.datacenter[].heap`, `cassandra.datacenter[].heap.size`, and `cassandra.datacenter[].heap.newGenSize`?

That seems pretty undesirable to me. I'd rather see a fix that modifies the template to make the right decision in the _presence_ of the empty settings. Would that be difficult? (Note that I'm not suggesting we calculate anything -- just do what it already does today.)

Reply:
> Am I understanding correctly that, due to these changes, our automatically generated settings docs will not contain specific entires for any of cassandra.heap, cassandra.heap.size, cassandra.heap.newGenSize, cassandra.datacenter[].heap, cassandra.datacenter[].heap.size, and cassandra.datacenter[].heap.newGenSize?

This is correct (at least in my testing), if the field is commented out the tooling doesn't see something that needs to be annotated.

Reply:
> That seems pretty undesirable to me. I'd rather see a fix that modifies the template to make the right decision in the presence of the empty settings. Would that be difficult? (Note that I'm not suggesting we calculate anything -- just do what it already does today.)

I couldn't agree with this more. That's my desired behavior.

Reply:
@bradfordcp @jakerobb @jsanda To clarify, we now want the properties to be existing at both the `cluster` and `datacenter` levels to support the `documentation generator` so our docs illustrate the` optional jvm heap` settings.

This was not the [original requirement](https://github.com/k8ssandra/k8ssandra/pull/242#discussion_r564590891) prior to this fix, which is why there is some confusion on my part.

My opinion at this point, I would like to see the heap config keys listed (without default values) available knowing that for now (v 1.0), the startup may fail due to unspecified values.  This is why I opened this item #348 today to consider post-1.0 and catch these using linting rules.  Once something like #348 were in place, the issue of missing values and/or missing configuration keys could be identified early on with some clarity.

Reply:
Let me provide a bit of background. I went back to look at the initial PR #242 to see what was discussed. We decided that we did not want to set defaults, i.e., non-empty values. There is a comment in which I suggest commenting them out after deciding we did not want to set defaults.

I also would like to see generated docs for all of the properties. With that said, if commenting out some or all of the properties make the code easier/simpler to development and maintain, then I will take easier/simpler code.

I have looked at some other chart projects for comparison. I see a common pattern for complex properties, i.e., having subproperties. The top-level property is set to an empty map or array. Comments are included to describe the subproperties. Here are a couples examples for reference:

* https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/values.yaml#L523
* https://github.com/bitnami/charts/blob/master/bitnami/metrics-server/values.yaml#L198

In general I think we can work with initializing the top-level property to an empty map unless it makes sense to set default values for nested properties. With this approach the doc tooling will pick up the top-level property. We can include comments with examples of nested properties. It also simplifies code because we can avoid having to check any number of nested properties.

@jeffbanks wdyt?

Reply:
@jeffbanks You bring up a good point about DC-level properties. I would do the same. Set the DC-level `heap` to an empty map.

I have not yet looked at #348 but post-1.0 I want to really look at moving more of the implementation out of Helm charts and into CRDs+controllers.

Reply:
@jsanda @bradfordcp 

I like that we would be consistent with others for management of `complex properties`.  As such, for k8ssandra I see the changes scoped to release 1.0 being the following:

1) `values.yaml` updated to include top-level heap property (at cluster and dc levels) with empty map value.
The heap's child properties will be commented out w/ no default values.

```yaml
cassandra:
  # cluster-level settings
  heap: {}
    # size:
    # newGenSize:
  # dc specific settings
  dataCenters:
    - name: dc1
      size: 1
      heap: {}
        # size:
        # newGenSize:
    - name: dc2
      size: 1
```

2) The comment format will follow the other comment entries to avoid any issues with the doc generator rules.  That is, usage of the '--' prefix.  This would resolve issues pointed out by @jakerobb in the generated docs.

3) Unit tests will be adjusted to allow for empty `heap` property values and with or without specified child properties at both levels.

4) If the k8ssandra user doesn't configure the heap settings, there are no default values provided.